astropy/io/fits/file.py

```diff
@@ -487,22 +521,7 @@ def _open_filename(self, filename, mode, overwrite):
 
```

Comment:
You could move the magic bytes reading here in the new method (`_test_for_compression`).